crates/aptos-compression/src/lib.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::metrics::{
+    increment_compression_byte_count, increment_compression_error,
+    start_compression_operation_timer, COMPRESS, COMPRESSED_BYTES, DECOMPRESS, RAW_BYTES,
+};
+use aptos_logger::prelude::*;
+use lz4::block::CompressionMode;
+use thiserror::Error;
+
+/// This crate provides a simple library interface for data compression.
+/// It is useful for compressing large data chunks that are
+/// sent across the network (e.g., by state sync and consensus).
+/// Internally, it uses LZ4 in fast mode to compress the data.
+/// See https://github.com/10xGenomics/lz4-rs for more information.
+///
+/// Note: the crate also exposes some basic compression metrics
+/// that can be used to track the cumulative compression ratio
+/// and compression/decompression durations during the runtime.
+mod metrics;
+#[cfg(test)]
+mod tests;
+
+/// The acceleration parameter to use for FAST compression mode.
+/// This was determined anecdotally.
+const ACCELERATION_PARAMETER: i32 = 1;
+
+/// A useful wrapper for representing compressed data
+pub type CompressedData = Vec<u8>;
+
+/// An error type for capturing compression/decompression failures
+#[derive(Clone, Debug, Error)]
+#[error("Encountered a compression error! Error: {0}")]
+pub struct CompressionError(String);
+
+/// Compresses the raw data stream
+pub fn compress(raw_data: Vec<u8>) -> Result<CompressedData, CompressionError> {
+    // Start the compression timer
+    let timer = start_compression_operation_timer(COMPRESS);
+
+    // Compress the data
+    let compression_mode = CompressionMode::FAST(ACCELERATION_PARAMETER);
```

Comment:
How do we feel about making this configurable, so that we can play around with different types of compression modes for different components?

Reply:
We could do this, but to be honest, I don't see the immediate need right now, e.g.,
- Non fast mode is very slow (10+ ms, without a significantly huge compression improvement).
- The `ACCELERATION_PARAMETER` doesn't make that much difference in terms of time/compression ratio.
- Other compression libraries haven't performed as well as this one has (from my own biased testing 😄).
- I don't really see a use case for this today (just yet). We also need to be aware that the client needs to have some control (e.g., it should determine the supported compression algorithms, etc.)